### ISSUE TITLE
Inplace fused (leaky-)relu+dropout for memory savings

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#92927]
### 🚀 The feature, motivation and pitch

It's possible that such an inplace fusion is already possible with dynamo, but if not, it's quite good to have for saving memory. Also maybe worth introducing some torch.nn.fused namespace which would have torch.nn.fused.ReLU_Dropout, torch.nn.fused.Conv2d_BatchNorm2d or some other similar modules that in the ultimate case might be defined as `torch.compile(nn.Sequential(nn.ReLU(inplace=True), nn.Dropout(inplace=True)))`, but in the meanwhile can be hand-optimized for these very frequent cases

My earlier issue in torchvision:
https://github.com/pytorch/vision/issues/4851

**Comment by ezyang** [comment]
I think Inductor is supposed to know it can reuse the buffer here but I am not sure the derivative formulas know though. Cc @Chillee

**Comment by vadimkantorov** [comment]
Reusing the buffer meaning inplace? Or something in between? Because normally, the input buffer needs to be saved for backward, and that's why it can't be reused except reversible cases and some other cases. Normally, autograd can't compose inplace relu and inplace dropout because of version tracking. But in the case ReLU + Dropout inplace fusion this is fine.

**Comment by Chillee** [comment]
@vadimkantorov For dropout + relu, we do this (run this with `AOT_FX_GRAPHS=1`)
```
@torch.compile
def f(x):
    return F.dropout(x, 0.5, training=True).relu()

f(torch.randn(20, 20, requires_grad=True, device='cuda'))
```
```
====== Forward graph 0 ======
class GraphModule(torch.nn.Module):
    def forward(self, primals_1: f32[20, 20]):
        # No stacktrace found for following nodes
        philox_seed_like: i32[] = torch.ops.prims.philox_seed_like.default(primals_1)
        philox_rand_like: f32[20, 20] = torch.ops.prims.philox_rand_like.default(primals_1, philox_seed_like, 0)
        gt: b8[20, 20] = torch.ops.aten.gt.Scalar(philox_rand_like, 0.5);  philox_rand_like = None
        convert_element_type: f32[20, 20] = torch.ops.prims.convert_element_type.default(gt, torch.float32);  gt = None
        mul: f32[20, 20] = torch.ops.aten.mul.Tensor(convert_element_type, primals_1);  convert_element_type = primals_1 = None
        mul_1: f32[20, 20] = torch.ops.aten.mul.Tensor(mul, 2.0);  mul = None
        
        # File: t.py:10, code: return F.dropout(x, 0.5, training=True).relu()
        relu: f32[20, 20] = torch.ops.aten.relu.default(mul_1);  mul_1 = None
        le: b8[20, 20] = torch.ops.aten.le.Scalar(relu, 0)
        return [relu, philox_seed_like, le]
        
====== Backward graph 0 ======
class GraphModule(torch.nn.Module):
    def forward(self, philox_seed_like: i32[], le: b8[20, 20], tangents_1: f32[20, 20]):
        # File: t.py:10, code: return F.dropout(x, 0.5, training=True).relu()
        scalar_tensor: f32[] = torch.ops.aten.scalar_tensor.default(0.0, dtype = torch.float32, layout = torch.strided, device = device(type='cpu'))
        where: f32[20, 20] = torch.ops.aten.where.self(le, scalar_tensor, tangents_1);  le = scalar_tensor = tangents_1 = None
        
        # File: t.py:10, code: return F.dropout(x, 0.5, training=True).relu()
        philox_rand_like_1: f32[20, 20] = torch.ops.prims.philox_rand_like.default(where, philox_seed_like, 0);  philox_seed_like = None
        gt_1: b8[20, 20] = torch.ops.aten.gt.Scalar(philox_rand_like_1, 0.5);  philox_rand_like_1 = None
        convert_element_type_1: f32[20, 20] = torch.ops.prims.convert_element_type.default(gt_1, torch.float32);  gt_1 = None
        mul_2: f32[20, 20] = torch.ops.aten.mul.Tensor(convert_element_type_1, where);  convert_element_type_1 = where = None
        mul_3: f32[20, 20] = torch.ops.aten.mul.Tensor(mul_2, 2.0);  mul_2 = None
        return [mul_3]
```

So we only save a single value (the random seed) + the relu mask, which seems pretty good to me?

**Comment by ngimel** [comment]
Pointwise ops are pretty reliably optimized by inductor, and new fused variants won't be added to pytorch core.

**Comment by vadimkantorov** [comment]
> So we only save a single value (the random seed) + the relu mask, which seems pretty good to me?

Can't it be completely inplace? (e.g. not even storing the relu (?) mask)

**Comment by vadimkantorov** [comment]
> Pointwise ops are pretty reliably optimized by inductor, and new fused variants won't be added to pytorch core.

Then maybe it's a good set of ops to showcase usefulness of dynamo/inductor. Another interesting direction is being able to somehow express in arguments to torch.compile that the result should be fully inplace / consume this many copies (or otherwise throw some error), then one can trust more the compiler

**Comment by vadimkantorov** [comment]
Can't the mask be recovered again from the torch.sign applied to saved_output?

**Comment by Chillee** [comment]
> Can't the mask be recovered again from the torch.sign applied to saved_output?

You could have a gradient formula that called torch.sign on saved_output, but that would require saving the output of `torch.relu` and reading it back in, which is more expensive for fp32 than just saving the mask of relu as a boolean array.

> And also it seems that relu call produced by inductor isn't inplace.

It's better than inplace - it's fused into a single kernel :) The 2 FX graphs I posted above result in 2 kernels total - one for the forwards pass and one for the backwards pass.

**Comment by vadimkantorov** [comment]
> It's better than inplace - it's fused into a single kernel :)

I think it can be both inplace and single kernel - saving even more memory! For these partial small fused blocks, somehow (optionally?) enabling inplace write-outs should be a good feature.

This is how I implemented it manually:
```python
def relu_dropout(x, p = 0, inplace = False, training = False):
    if not training or p == 0:
        return x.clamp_(min = 0) if inplace else x.clamp(min = 0)

    p1m = 1 - p
    mask = torch.rand_like(x) > p1m
    # mask = torch.bernoulli(torch.tensor(p, device = x.device, dtype = torch.float32).expand_as(x))
    mask |= (x < 0) # https://github.com/pytorch/pytorch/issues/41793
    return x.masked_fill_(mask, 0).div_(p1m) if inplace else x.masked_fill(mask, 0).div(p1m)
```

If such a forward can be fused (and even better if it can be created automatically from `nn.Sequential(nn.ReLU(inplace = True), nn.Dropout(inplace = True))` and made inplace - that would be very nice.

For very frequent nets with blocks like Conv+Bn+(Leaky)Relu+Dropout this is a big win in memory, as all can be done in reversible fashion and inplace

**Comment by Chillee** [comment]
@vadimkantorov At least from my benchmarks, I think torch.compile will just give as good/better results as your manual impl?
```
import torch
import torch.nn.functional as F

def bench(f, name=None, iters=100, warmup=5, display=True, profile=False):
    import time
    for _ in range(warmup):
        f()
    if profile:
        with torch.profiler.profile() as prof:
            f()
        prof.export_chrome_trace(f"{name if name is not None else 'trace'}.json")

    torch.cuda.synchronize()
    begin = time.time()
    for _ in range(iters):
        f()
    torch.cuda.synchronize()
    us_per_iter = (time.time()-begin)*1e6/iters

    if name is None:
        res = us_per_iter
    else:
        res= f"{name}: {us_per_iter:.3f}us"

    if display:
        print(res)
    return res

def custom_relu_dropout(x, p = 0, inplace = False, training = False):
    if not training or p == 0:
        return x.clamp_(min = 0) if inplace else x.clamp(min = 0)

    p1m = 1 - p
    mask = torch.rand_like(x) > p1m
    # mask = torch.bernoulli(torch.tensor(p, device = x.device, dtype = torch.float32).expand_as(x))
    mask |= (x < 0) # https://github.com/pytorch/pytorch/issues/41793
    return x.masked_fill_(mask, 0).div_(p1m) if inplace else x.masked_fill(mask, 0).div(p1m)

def relu_dropout(x, p=0, training=True):
    return F.dropout(torch.relu(x), p=p, training=training)

inp = torch.randn(2**26, device='cuda', requires_grad=True)

compiled_custom = torch.compile(custom_relu_dropout)
compiled_relu = torch.compile(relu_dropout)
val = custom_relu_dropout(inp, p=0.5, training=True)

print(torch.cuda.memory_allocated()/1e9)
del val

val = relu_dropout(inp, p=0.5, training=True)
print(torch.cuda.memory_allocated()/1e9)
del val

val = compiled_custom(inp, p=0.5, training=True)
print(torch.cuda.memory_allocated()/1e9)
del val

val = compiled_relu(inp, p=0.5, training=True)
print(torch.cuda.memory_allocated()/1e9)
del val


bench(lambda: custom_relu_dropout(inp, p=0.5, training=True).sum().backward(), "custom eager")
bench(lambda: relu_dropout(inp, p=0.5, training=True).sum().backward(), "normal eager")

bench(lambda: compiled_custom(inp, p=0.5, training=True).sum().backward(), "custom compiled")
bench(lambda: compiled_relu(inp, p=0.5, training=True), "normal compiled")
```
```
Memory used:
custom eager: 0.603979776GB
normal eager: 0.872415232
custom compiled: 0.603980288
normal compiled: 0.603981312

custom eager: 3989.549us
normal eager: 2585.723us
custom compiled: 1932.619us
normal compiled: 461.659us
```

**Comment by vadimkantorov** [comment]
it's maybe because my script itself isn't optimal :) it seems that the whole thing can't be made allocation-less :)

**Comment by ngimel** [comment]
High water mark of memory is achieved towards the end of the forward pass and is determined by the things one needs to save for backward. Making some intermediate operations inplace instead of outplace has no to minimal effect on high watermark (e.g. you can skip allocating the output and modify input inplace if you don't need it for backward, but then the input will be released and reused in the next op), so here we are talking about microoptimizations with questionable e2e benefits. If you feel those are critical you can write custom ops for those cases.

**Comment by vadimkantorov** [comment]
I've experimented on inplacing BatchNorm for JasperNet-like models, the result was reduction of 10-20% of CUDA mem usage. I agree that in general, these are often too brittle (and that for benefits to be realizable, the full network better be reversible except some specific cases). My idea is that: 
- there're some specific sets of blocks that are already frequent in practice for quite some years that might warrant inclusion/discussion of fused+inplace variants
- if not inclusion, it can be discussed that torch.compile be controlled to produce inplace kernels if asked so (if for instance, it could not compile `nn.Sequential(nn.ReLU(inplace=True), nn.Dropout(inplace=True))`)

**Comment by vadimkantorov** [comment]
btw if torch.compile can produce inplace code, it could be used for compiling optimizers code and maybe generating foreach variants automatically

**Comment by vadimkantorov** [comment]
> btw if torch.compile can produce fully inplace code, it could be used for compiling optimizers code and maybe generating foreach variants automatically

E.g. one could then implement simpler something like mix3 from https://github.com/pytorch/pytorch/issues/71683#issuecomment-1379058082

**Comment by albanD** [comment]
Custom autograd Functions should already work today :)

**Comment by vadimkantorov** [comment]
one can just add @torch.compile to `forward` and `backward` methods of the `torch.autograd.Function` subclass?

**Comment by albanD** [comment]
Either that or add it to whatever calls `YourFn.apply()`

**Comment by vadimkantorov** [comment]
Hmm, I also wonder if swish+dropout(==`x * sigmoid(x) * mask_of_zeros_and_dropout_multiples` can be fused/done inplace during training?

This combination is occurring in FFN of Conformer blocks and can save some memory: https://github.com/NVIDIA/NeMo/blob/07b79d2/nemo/collections/asr/parts/submodules/conformer_modules.py#L383

**Comment by williamwen42** [comment]
Can we close this issue? Does `torch.compile` perform the desired fusion today?

**Comment by vadimkantorov** [comment]
I'll do a new test! Can torch.compile currently produce a truly inplace code which would result in no copies/allocations completely? 

This is about compiling a manually fused impl.

Regarding a more general case, I wonder if this fusion can be implemented as a compiler pass? (e.g. producing fully allocation-less code for `nn.Sequential(nn.ReLU(inplace=True), nn.Dropout(inplace=True))`)

Ideally, Conv+BatchNorm+Relu+Dropout+Residual is a quite common pattern in convnets

**Comment by vadimkantorov** [comment]
A bit related:
- https://github.com/pytorch/pytorch/issues/135994

I think it would be best to include some of these patterns as "known shortcuts" with regression tests for torch.compile options (maybe at least have these patterns enumerated somewhere in docs)...

**Comment by vadimkantorov** [comment]
Might be related:
- https://github.com/pytorch/pytorch/issues/158643
